Results_Engine/Query/MapResults.cs

```diff
@@ -111,6 +113,89 @@ private static string IdMatch(this IBHoMObject o, Type identifier)
+                    Engine.Base.Compute.RecordNote("No identifier found or specified. BHoM_Guid will be used to identify the Object.");
+                    return x => x.BHoM_Guid.ToString();
+                }
+            }
+            else if(identifier is Type)
+            {
+                //If identifier type provided is Type, check that the type is valid
+                identifier = obj.FindIdentifier(identifier as Type);
+                if (identifier != null)
+                {
+                    Type typeId = identifier as Type;
+                    return x => IdMatch(x, typeId);
+                }
+                else
+                    return null;
+            }
+            else if (identifier is string)
+            {
+                //If string
+
+                //Check if string is a type string in a BHoM namespace
+                if ((identifier as string).StartsWith("BH."))
+                {
+                    Type type = Create.Type(identifier as string, true);
+                    if (type != null)
+                        return GetObjectIdentifier(obj, type);
+                }
+
+                string idString = (identifier as string).ToLower();
+
+                //Check if name or Guid. If so return property extractor as optimisation
+                if (idString == "name")
+                    return x => x.Name;
+
+                if (idString == "bhom_guid" || idString == "guid")
+                    return x => x.BHoM_Guid.ToString();
+
+                //If not, rely on the slower running but more generic PropertyValue
+                return x => Base.Query.PropertyValue(x, identifier as string).ToString();
+            }
+
+
+            Engine.Base.Compute.RecordError("Identifier should either be a IAdapterId type or a string corresponding to the property of the object used to be used to identify the object.");
+            return null;
+        }
+
+        /***************************************************/
+
+        private static Func<T, string> GetResultIdentifier<T>(this T obj, string identifier) where T : IResult
+        {
+            //If result type and identifier match for ObjectIdResult or MeshResult, create custom lamdas to speed things up
+            string idLower = identifier.ToLower();
+            if (idLower == nameof(IObjectIdResult.ObjectId).ToLower() && obj is IObjectIdResult)
+                return x => ((IObjectIdResult)x).ObjectId.ToString();
+
+            if (idLower == nameof(IMeshElementResult.NodeId).ToLower() && obj is IMeshElementResult)
+                return x => ((IMeshElementResult)x).NodeId.ToString();
+
+            if (idLower == nameof(IMeshElementResult.MeshFaceId).ToLower() && obj is IMeshElementResult)
+                return x => ((IMeshElementResult)x).MeshFaceId.ToString();
+
+            //Fall back on PropertyValue
+            return x => Base.Query.PropertyValue(x, identifier).ToString();
+        }
```

Comment:
Would be nice to make available publicly as per comment https://github.com/BHoM/BHoM_Engine/pull/2768/files#r822565874